### PR TITLE
Fixes issue with code generation of net_send_buffering()

### PR DESCRIPTION
**Pull request by iomaganaris**
- If `OpenMP` GPU backend is enabled then `#pragma omp atomic capture` is not executed properly on CPU with NVHPC 23.1 (more details and tests in https://github.com/neuronsimulator/nrn/pull/2239)
- Wrap the update of `nsb->_cnt` inside an `if-statement` based on `nt->compute_gpu`

### CODE REVIEW

**Comment by pramodk** [comment]
> #pragma omp atomic capture is not executed properly on CPU with NVHPC 23.1

does this produce bogus results? 🤔

**Comment by iomaganaris** [comment]
> > #pragma omp atomic capture is not executed properly on CPU with NVHPC 23.1
> 
> does this produce bogus results? thinking

Yes, without these changes on CPU we get:
```
➜  coreneuron git:(magkanar/fix_nvhpc_231) ✗ ./x86_64/special -mpi -python test_watchrange.py         
numprocs=1
NEURON -- VERSION 9.0.dev-1300-g8d5f19d39+ magkanar/fix_nvhpc_231 (8d5f19d39+) 2023-02-20
Duke, Yale, and the BlueBrain Project -- Copyright 1984-2022
See http://neuron.yale.edu/neuron/credits

Additional mechanisms from files
 "mod files/axial.mod" "mod files/axial_pp.mod" "mod files/bacur.mod" "mod files/banocur.mod" "mod files/fornetcon.mod" "mod files/invlfire.mod" "mod files/netmove.mod" "mod files/sample.mod" "mod files/watchrange.mod"
CoreNEURON run
==== 0 ====
==== 1 ====
==== 2 ====
net_sem_from_gpu sendtype 0 i_vdata 2 weight_index_ -1 ith 0 ipnt 0 td 0.000000 flag 3.000000
net_sem_from_gpu sendtype 0 i_vdata 17 weight_index_ -1 ith 0 ipnt 15 td 0.000000 flag 3.000000
net_sem_from_gpu sendtype 0 i_vdata 26 weight_index_ -1 ith 0 ipnt 24 td 0.000000 flag 1.000000
==== 34 ====
==== 35 ====
==== 36 ====
net_sem_from_gpu sendtype 0 i_vdata 2 weight_index_ -1 ith 0 ipnt 0 td 0.000000 flag 3.000000
net_send td-t = -0.125 SelfEvent target=Bounce 0 flag=3
[magkanar-MS-7C73:174349] *** Process received signal ***
[magkanar-MS-7C73:174349] Signal: Aborted (6)
[magkanar-MS-7C73:174349] Signal code:  (-6)
[magkanar-MS-7C73:174349] [ 0] /usr/lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7f62ba442520]
[magkanar-MS-7C73:174349] [ 1] /usr/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7f62ba496a7c]
[magkanar-MS-7C73:174349] [ 2] /usr/lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7f62ba442476]
[magkanar-MS-7C73:174349] [ 3] /usr/lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7f62ba4287f3]
[magkanar-MS-7C73:174349] [ 4] /home/magkanar/bbp_repos/nrn/test/coreneuron/x86_64/libcorenrnmech.so(+0x12774e)[0x7f62be12774e]
[magkanar-MS-7C73:174349] [ 5] /home/magkanar/bbp_repos/nrn/test/coreneuron/x86_64/libcorenrnmech.so(_ZN10coreneuron22net_buf_receive_BounceEPNS_9NrnThreadE+0x49b)[0x7f62be0998db]
```

With these changes on CPU (and same on GPU):
```
➜  coreneuron git:(magkanar/fix_nvhpc_231) ✗ ./x86_64/special -mpi -python test_watchrange.py         
numprocs=1
NEURON -- VERSION 9.0.dev-1300-g8d5f19d39+ magkanar/fix_nvhpc_231 (8d5f19d39+) 2023-02-20
Duke, Yale, and the BlueBrain Project -- Copyright 1984-2022
See http://neuron.yale.edu/neuron/credits

Additional mechanisms from files
 "mod files/axial.mod" "mod files/axial_pp.mod" "mod files/bacur.mod" "mod files/banocur.mod" "mod files/fornetcon.mod" "mod files/invlfire.mod" "mod files/netmove.mod" "mod files/sample.mod" "mod files/watchrange.mod"
CoreNEURON run
==== 0 ====
==== 1 ====
==== 2 ====
net_sem_from_gpu sendtype 0 i_vdata 26 weight_index_ -1 ith 0 ipnt 24 td 0.000000 flag 1.000000
net_sem_from_gpu sendtype 0 i_vdata 17 weight_index_ -1 ith 0 ipnt 15 td 0.000000 flag 3.000000
net_sem_from_gpu sendtype 0 i_vdata 2 weight_index_ -1 ith 0 ipnt 0 td 0.000000 flag 3.000000
==== 0 ====
==== 1 ====
==== 2 ====
```

File: src/codegen/codegen_acc_visitor.cpp [hunk 1]
```diff
@@ -120,6 +120,16 @@ void CodegenAccVisitor::print_abort_routine() const {
     printer->end_block(1);
 }
 
+void CodegenAccVisitor::print_net_send_buffering_cnt_update() const {
+    printer->fmt_start_block("if (nt->compute_gpu)");
+    print_device_atomic_capture_annotation();
+    printer->add_line("i = nsb->_cnt++;");
+    printer->decrease_indent();
+    printer->start_block("} else");
```
Comment (olupton) [inline src/codegen/codegen_acc_visitor.cpp:128]:
```suggestion
    printer->restart_block("else");
```

File: src/codegen/codegen_cpp_visitor.cpp [hunk 2]
```diff
@@ -3714,8 +3714,8 @@ void CodegenCVisitor::print_watch_check() {
         printer->add_indent();
         printer->add_text("net_send_buffering(");
         auto t = get_variable_name("t");
-        printer->add_text(
-            fmt::format("ml->_net_send_buffer, 0, {}, -1, {}, {}+0.0, ", tqitem, point_process, t));
+        printer->add_text(fmt::format(
+            "nt, ml->_net_send_buffer, 0, {}, -1, {}, {}+0.0, ", tqitem, point_process, t));
```
Comment (olupton) [inline src/codegen/codegen_cpp_visitor.cpp:3718]:
```suggestion
        printer->fmt_text("nt, ml->_net_send_buffer, 0, {}, -1, {}, {}+0.0, ", tqitem, point_process, t);
```

**Comment by iomaganaris** [comment]
I'm a bit skeptical about the fact that coverage of [src/codegen/codegen_acc_visitor.cpp](https://codecov.io/gh/BlueBrain/nmodl/pull/1005?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=BlueBrain#diff-c3JjL2NvZGVnZW4vY29kZWdlbl9hY2NfdmlzaXRvci5jcHA=) increased so much while testing only a very small part of it but I guess it's better than reducing it. We can follow up on that in https://github.com/BlueBrain/nmodl/issues/957

**Review by olupton — COMMENTED** [review]
I'm curious...are we sure _all_ these changes to the code generation needed?
Did you check if the change to `print_net_send_buf_update_to_host` is sufficient without the changes to the atomic update part?

A couple of code-printing nitpicks.

**Comment by iomaganaris** [comment]
> I'm curious...are we sure _all_ these changes to the code generation needed? Did you check if the change to `print_net_send_buf_update_to_host` is sufficient without the changes to the atomic update part?
> 
> A couple of code-printing nitpicks.

Yes, the changes to `print_net_send_buf_update_to_host` are not enough. I just included them since I've already done them. I can try to create a small reproducer of the issue as well to report it to NVIDIA

**Review by olupton — APPROVED** [review]

**Comment by iomaganaris** [comment]
> > Use C++ `if-statement` instead of in OpenMP pragma in some cases
> 
> Bit late here - @iomaganaris: similar to the two below bug reports, did you see if it's worth reporting to nvofurm? (if not already reported)
> 
> https://forums.developer.nvidia.com/t/incorrect-gpu-results-with-pragma-acc-atomic-capture/187962 https://forums.developer.nvidia.com/t/incorrect-cpu-results-with-pragma-acc-atomic-capture/184453/

I am trying to create a small reproducer but I haven't managed until now. I was just able to reproduce the first issue of these with my own example. I will try with Olli's as well.

> Use C++ `if-statement` instead of in OpenMP pragma in some cases
Note that this was not needed for the fix but I find it a bit cleaner and maybe less error prone.

**Pull request merged**